### PR TITLE
feat: support mulit/range

### PR DESCRIPTION
**Pull request by byene0923**
pr for https://github.com/valyala/fasthttp/issues/1312

- change `SetByteRange`, support multipart/byteranges
- change `bigFileReader` and `fsSmallFileReader` which only handle complete files
- add `smallRangeReader` and `bigRangeReader` which only handle range files
- change `byteRangeUpdater` interface to `byteRangeHandler` 
**test**
- change `TestFSByteRangeSingleThread`, add `bigFileReader` and `fsSmallFileReader` test of `read` and `WriteTo`
- add `TestFSMultiByteRangeConcurrent` and `TestFSMultiByteRangeSingleThread`
- change `TestParseByteMultiRangeSuccess` and `TestParseByteRangeError`

### CODE REVIEW

File: fs.go [hunk 1]
```diff
@@ -935,25 +977,29 @@ func (h *fsHandler) handleRequest(ctx *RequestCtx) {
 	statusCode := StatusOK
 	contentLength := ff.contentLength
 	if h.acceptByteRange {
-		hdr.setNonSpecial(strAcceptRanges, strBytes)
+		hdr.SetCanonical(strAcceptRanges, strBytes)
```
Comment (Aoang) [inline fs.go:980]: If `fsSmallFileReader` does not support `ByteRange`, cannot set `Accept-Ranges: bytes`.

File: header.go [hunk 2]
```diff
@@ -113,18 +113,23 @@ func (h *ResponseHeader) SetContentRange(startPos, endPos, contentLength int) {
 //
 //   - If startPos is negative, then 'bytes=-startPos' value is set.
 //   - If endPos is negative, then 'bytes=startPos-' value is set.
-func (h *RequestHeader) SetByteRange(startPos, endPos int) {
+func (h *RequestHeader) SetByteRange(startPos, endPos []int) {
```
Comment (erikdubbelboer) [inline header.go:116]: This is a backwards compatible change we can not make.

File: header_test.go [hunk 3]
```diff
@@ -2922,4 +2922,4 @@ func expectResponseHeaderAll(t *testing.T, h *ResponseHeader, key string, expect
 	if !reflect.DeepEqual(h.PeekAll(key), expectedValue) {
 		t.Fatalf("Unexpected value for key %q: %q. Expected %q", key, h.PeekAll(key), expectedValue)
 	}
-}
+}
```
Comment (erikdubbelboer) [inline header_test.go:2925]: Why are you removing all the newlines at the end of files? They should be there.

**Review by erikdubbelboer — CHANGES_REQUESTED** [review]

File: header_test.go (hunk 3 above)
Comment (byene0923) [reply header_test.go:2925]: I use goland ide, the format is different from github lint, i will fix it soon

File: header.go (hunk 2 above)
Comment (byene0923) [reply header.go:116]: yeah, add  method of  `SetByteRanges`?

**Review by erikdubbelboer — CHANGES_REQUESTED** [review]

File: fs.go [hunk 4]
```diff
@@ -607,143 +610,175 @@ func (ff *fsFile) decReadersCount() {
...
+	return ff.smallRangeReader(), nil
+}
+
+func (ff *fsFile) smallRangeReader() io.Reader {
+	v := ff.h.smallRangeReaderPool.Get()
+	if v == nil {
+		v = &smallRangeReader{
+			sta: make([]int, 64),
+			end: make([]int, 64),
+			buf: make([]byte, 4096),
```
[the remark this answers is not in the document]
Comment (byene0923) [reply fs.go:630]: 4096 keep the same with [`copyBufPool`](https://github.com/valyala/fasthttp/blob/master/http.go#L2052), this use for `copyZeroAlloc`

File: fs.go [hunk 5]
```diff
@@ -1451,3 +1498,499 @@ func getFileLock(absPath string) *sync.Mutex {
...
+		cPos += n
+		r.buf = r.buf[:0]
+	}
+	return cPos, io.EOF
+}
+
+func (r *bigRangeReader) WriteTo(w io.Writer) (int64, error) {
+	if rf, ok := w.(io.ReaderFrom); ok {
+		return rf.ReadFrom(r)
+	}
```
[the remark this answers is not in the document]
Comment (byene0923) [reply fs.go:1888]: this is for `bytebufferpool`, do as this 
```
bb := bytebufferpool.Get()
_, err := reader.WriteTo(bb)
```
and `smallRangeReader` has also do it, in fs.go  line 1690. `smallRangeReader` do the same as `smallFileReader`, handle directory first, then handle `readfrom`, last handle `write`

File: fs.go [hunk 6]
```diff
@@ -1451,3 +1498,499 @@ func getFileLock(absPath string) *sync.Mutex {
...
+		}
+
+		if r.IsMultiRange() && !r.hasCurRangeBodyHeader {
+			r.hasCurRangeBodyHeader = true
+			multiRangeBodyHeader(&r.buf, r.sta[i], r.end[i], r.ff.contentLength, r.ff.contentType, r.boundary, r.hf)
+			r.hf = false
+			cLen = len(p[cPos:])
+			n = copy(p[cPos:], r.buf)
+			cPos += n
+			if len(r.buf) > cLen {
```
[the remark this answers is not in the document]
Comment (byene0923) [reply fs.go:1593]: this is a bit complex
the multibyte/range like
```
--3d6b6a416f9b5
Content-Type: text/html
Content-Range: bytes 0-15/1270

<!DOCTYPE html>
--3d6b6a416f9b5
Content-Type: text/html
Content-Range: bytes 0-10/1270

<!DOCTYPE 
--3d6b6a416f9b5--
```
for each range, we need to add `boundary`, `Content-Type`, `Content-Range` first, then add http content, at last range, we need to and `boundary--`

however, the length of `p` is limited, so we need to check whether `p` is full when we add `multiRangeBodyHeader` which contain `boundary`, `Content-Type`, `Content-Range` and http content, and `multiRangeBodyEnd` which contain `boundary--`

the `r.buf` is the rest content of `multiRangeBodyHeader` or `multiRangeBodyEnd` that has not copy into `p` at last time.
`cLen` is the rest lengh of `p`,  and we have done `n = copy(p[cPos:], r.buf)`, if `cLen` is less then `len(r.buf)`, we can copy all the content of `r.buf` to `p`

by the way,
method `read` like a streaming which the `p` is limited
first, we read the `multiRangeBodyHeader`, then we read http content, if we are in last range, we read the `multiRangeBodyEnd`

the code which handle range content begins  in line fs.go 1603

**Review by erikdubbelboer — CHANGES_REQUESTED** [review]
It seems like `TestFSMultiByteRangeConcurrent` always gets a timeout, can you fix that any other tests that might fail?

**Comment by byene0923** [comment]
> 
sure,  the  test number of concurrent and multirange may be  too large

**Comment by erikdubbelboer** [comment]
Thanks!

**Comment by erikdubbelboer** [comment]
@byene0923 I had to revert this change as it caused a critical bug: https://github.com/valyala/fasthttp/issues/1445

Can you have a look at adding a test case for this and fixing it?

**Comment by byene0923** [comment]
ok, sorry for missing the full tests. i will look deep into it

**Comment by MindMayhem** [comment]
@byene0923  maybe it will be useful to you, but when we request large static file, then in request headers an interesting heading appears - Range: 3896-3896
on others files there's no such header

also, it's worth noting that my file was 5mb in size

**Comment by byene0923** [comment]
@MindMayhem thanks a lot, but i can't  reproduce this by your files in https://github.com/valyala/fasthttp/issues/1445. my system is  macOS 10.15.7 i will do it in different systems

**Comment by byene0923** [comment]
@MindMayhem thanks a lot ,but i can not reproduce it  by your files in https://github.com/valyala/fasthttp/issues/1445 now. should any changes i do in your code ?

**Comment by MindMayhem** [comment]
@byene0923 that's not my example, but i pretty sure it shows that error, just check your browser console
i hope @smalloff can provide more info to you

**Comment by erikdubbelboer** [comment]
I was able to reproduce it on MacOS using the files in the zipfile in this comment: https://github.com/valyala/fasthttp/issues/1445#issuecomment-1328800669 Only thing I did was `go mod tidy` and `go run main.go`.

**Comment by li-jin-gou** [comment]
> I was able to reproduce it on MacOS using the files in the zipfile in this comment: [#1445 (comment)](https://github.com/valyala/fasthttp/issues/1445#issuecomment-1328800669) Only thing I did was `go mod tidy` and `go run main.go`.

I'll try.😏

**Comment by byene0923** [comment]
thanks a lot to @MindMayhem and @erikdubbelboer, i reproduce it, i will look into it soon

**Comment by byene0923** [comment]
hello, @erikdubbelboer @MindMayhem i have found the bug in fasthttp@1.42.0 fs.go, line 749
```
func (r *fsSmallFileReader) WriteTo(w io.Writer) (int64, error) {
	if r.offset != 0 {
		panic("BUG: non-zero offset! Read() mustn't be called before WriteTo()")
	}
```
when file size  in (4096, 8192) will cause this bug.  because writeTo() will be called twice, and the second call `r.offset` is not zero, so this will cause panic, when file size > 8192, it use `bigFileReader`

many thanks to @MindMayhem  and @smalloff ! i will fix it soon and add unit tests

**Comment by byene0923** [comment]
it did a strange bug, and pass all the unit tests !

**Comment by byene0923** [comment]
i delete this check. and i'm doing the unit test, you could change the go.mod of the zipfile in this comment:  https://github.com/valyala/fasthttp/issues/1445#issuecomment-1328800669

just add this at the last line of go.mod

```
replace github.com/valyala/fasthttp => github.com/byene0923/fasthttp v1.23.1-0.20221202163206-7e40fb23488f
```
if @erikdubbelboer @MindMayhem have interest, can have a try

**Pull request merged**